### PR TITLE
[name] add removeName method to table__n_a_m_e

### PR DESCRIPTION
**Pull request by m4rc1e**
We can currently get and set names but we don't have a method to remove them. I've written similar code in several throwaway scripts before so I think it warrants its own method.

### CODE REVIEW

**Comment by anthrotype** [comment]
would also be nice to have a `removeNames` (plural) method to remove _any_ name record that matches given combination of nameID, platformID, platEncID or langID.
E.g. say one wants to drop all records with platformID=1.
The signature of `removeNames` could have all of arguments default to None 

```python
def removeNames(self, nameID=None, platformID=None, platEncID=None, langID=None):
```

the methods iterates over all name records and only keeps those which don't match the input
(no-op if all args are None).

File: Lib/fontTools/ttLib/tables/_n_a_m_e.py [hunk 1]
```diff
@@ -147,6 +147,18 @@ def setName(self, string, nameID, platformID, platEncID, langID):
...
+		"""Remove a name record identified by 'nameID', 'platformID', 'platEncID'
+                and 'langID'.
+		"""
+		for i, namerecord in enumerate(self.names):
+			if (namerecord.nameID == nameID and
+				namerecord.platformID == platformID and
+				namerecord.platEncID == platEncID and
+				namerecord.langID == langID):
+				    del self.names[i]
+				    break
```
Comment (anthrotype) [inline Lib/fontTools/ttLib/tables/_n_a_m_e.py]: how about
```python
self.names = [
    n for n in self.names
    if n.nameID != nameID
    and n.platformID != platformID
    and n.platEncID != platEncID
    and n.langID != langID
]
```

Comment (m4rc1e) [reply Lib/fontTools/ttLib/tables/_n_a_m_e.py]: This approach will definitely work better for removeNames. I'll update and include both methods.

**Comment by m4rc1e** [comment]
I've added the removeNames by using [functools.partialmethod](https://docs.python.org/3/library/functools.html#functools.partialmethod). Python 3.4+ is needed in order to use it. Happy to refactor if we need to support earlier versions of Python.

**Comment by m4rc1e** [comment]
Another approach without using partialmethod could be:

```
	def removeNames(self, nameID=None, platformID=None, platEncID=None, langID=None):
		self.names = [n for n in self.names
				if n.nameID != nameID
				if n.platformID != platformID
				if n.platEncID != platEncID
				if n.langID != langID
		]

        def removeName(self, nameID, platformID, platEncID, langID):
            self.removeNames(nameID, platformID, platEncID, langID)
```

**Comment by justvanrossum** [comment]
I think `removeNames` is enough. I'm not sure what the added benefit of `removeName` is to be honest.

**Comment by anthrotype** [comment]
I agree with Just, just removeNames is sufficient.

**Comment by m4rc1e** [comment]
Thanks for the feedback both of you. I've dropped `removeName` and kept `removesNames`.

File: Lib/fontTools/ttLib/tables/_n_a_m_e.py [hunk 2]
```diff
@@ -147,6 +147,17 @@ def setName(self, string, nameID, platformID, platEncID, langID):
 		else:
 			self.names.append(makeName(string, nameID, platformID, platEncID, langID))
 
+	def removeNames(self, nameID=None, platformID=None, platEncID=None, langID=None):
+		"""Remove name records identified by 'nameID', 'platformID', 'platEncID'
+		and 'langID'.
+		"""
+		self.names = [n for n in self.names
```
Comment (anthrotype) [inline Lib/fontTools/ttLib/tables/_n_a_m_e.py]: I prefer if add a newline after the opening `[`, with a single indentation level

**Comment by anthrotype** [comment]
please do `git fetch origin && git rebase origin/master` to get rid of unrelated extra commits

File: Lib/fontTools/ttLib/tables/_n_a_m_e.py [hunk 3]
```diff
@@ -147,6 +147,18 @@ def setName(self, string, nameID, platformID, platEncID, langID):
 		else:
 			self.names.append(makeName(string, nameID, platformID, platEncID, langID))
 
+	def removeNames(self, nameID=None, platformID=None, platEncID=None, langID=None):
+		"""Remove name records identified by 'nameID', 'platformID', 'platEncID'
+		and 'langID'.
```
Comment (anthrotype) [inline Lib/fontTools/ttLib/tables/_n_a_m_e.py]:
```suggestion
		or 'langID'.
```

File: Tests/ttLib/tables/_n_a_m_e_test.py [hunk 4]
```diff
@@ -75,6 +75,28 @@ def test_addName(self):
 		with self.assertRaises(TypeError):
 			table.addName(b"abc")  # must be unicode string
 
+	def test_removeNames(self):
```
Comment (anthrotype) [inline Tests/ttLib/tables/_n_a_m_e_test.py]: you could also add explicit tests for `removeNames` when called with more than one argument, especially one with all four of them.

File: Tests/ttLib/tables/_n_a_m_e_test.py [hunk 5]
```diff
@@ -75,6 +75,28 @@ def test_addName(self):
...
+		table = table__n_a_m_e()
+		table.setName("Regular", 2, 1, 0, 0)
+		table.setName("Regular", 2, 3, 1, 0.409)
```
Comment (anthrotype) [inline Tests/ttLib/tables/_n_a_m_e_test.py]: typo?

```suggestion
		table.setName("Regular", 2, 3, 1, 0.409)
		table.setName("Regular", 2, 3, 1, 0x409)
```

**Comment by anthrotype** [comment]
this should work:

```python
def removeNames(self, nameID=None, platformID=None, platEncID=None, langID=None):
    """Remove any name records identified by 'nameID', 'platformID', 'platEncID'
    or 'langID'.
    """
    args = {
        argName: argValue
        for argName, argValue in (
            ("nameID", nameID),
            ("platformID", platformID),
            ("platEncID", platEncID),
            ("langID", langID),
        )
        if argValue is not None
    }
    if not args:
        # no arguments, nothing to do
        return
    self.names = [
        rec for rec in self.names
        if any(
            argValue != getattr(rec, argName)
            for argName, argValue in args.items()
        )
    ]
```

**Comment by m4rc1e** [comment]
I came up with:

```
def removeNames(self, nameID=None, platformID=None, platEncID=None, langID=None):
        """Remove name records identified by 'nameID', 'platformID', 'platEncID'
        and 'langID'.
        """
        if not any ([nameID, platformID, platEncID, langID]):
                return
        nameIDs, platformIDs, platEncIDs, langIDs = set(), set(), set(), set()
        for name in self.names:
                if name.nameID == nameID or nameID is None:
                        nameIDs.add(name)
                if name.platformID == platformID or platformID is None:
                        platformIDs.add(name)
                if name.platEncID == platEncID or platEncID is None:
                        platEncIDs.add(name)
                if name.langID == langID or langID is None:
                        langIDs.add(name)
        to_remove = nameIDs & platformIDs & platEncIDs & langIDs
        self.names = sorted(set(self.names) - to_remove)
```

**Comment by m4rc1e** [comment]
Yours is cleaner so I'm going to use it :-). Thank you for your assistance on this.

**Comment by justvanrossum** [comment]
Why is Cosimo's original listcomp not good enough?

**Comment by anthrotype** [comment]
> Why is Cosimo's original listcomp not good enough?

because we want the method to drop name records when _all_ the given (non-None) parameters match. My simpler list comprehension was too greedy, as it dropped any records when _any_ of the input parameters match the corresponding attribute.

**Comment by justvanrossum** [comment]
> My simpler list comprehension was too greedy

Duh of course, now I see it. Thanks!

**Pull request merged**